### PR TITLE
[breaking] uniform cli commands and flag

### LINKED ISSUE #1508: Factor CLI argument processing for flags that are common across commands
We could use an approach similar to the `--port/-p, --protocol/-l` flags: https://github.com/arduino/arduino-cli/blob/master/cli/arguments/port.go

Flags that comes to mind:
* sketch path argument
* `--fqbn/-b` (sometimes is determined from `sketch.json` inside sketch path)

### PR DESCRIPTION
**Pull request by umbynos**
**Please check if the PR fulfills these requirements**

- [X] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [X] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] `UPGRADING.md` has been updated with a migration guide (for breaking changes)

* **What kind of change does this PR introduce?**

refactor CLI commands and flags to be consistent with each other. 
- **What is the current behavior?**

close  #1508
* **What is the new behavior?**

This PR includes various improvements:
  - sketchpath calculation have been factored out
  - fqbn and programmer flags now have the same principle of operation as port flag
  - factored out a function that reports errors when conflicting flags are used
  - move vars on top and remove structs for uniformity
  - part of the init section in `update-index` commands have been factored out
  - other minor improvements
  - I also added `logrus.Info()` in every run function of every command
  - remove `board details <fqbn>` in favour of `board details -b <fqbn>`
  - replace `board attach <port|fqbn> <sketch-path>` with `board attach -b <fqbn> | -p <port> <sketh-path>`
  - change `--timeout` flag to `--discovery-timeout` in `board list` command for consistency

- **Does this PR introduce a breaking change, and is
[titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?**

yep, documented in [`UPGRADING.md`](https://github.com/arduino/arduino-cli/blob/4d2df3155b190c2e0f6eca38c8ee07eb2c0b18f2/docs/UPGRADING.md)

* **Other information**:

---

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

### CODE REVIEW

File: cli/instance/instance.go
Comment (per1234) [inline cli/instance/instance.go:145]:
```suggestion
// It is mandatory for all `update-index` commands to call this
```

**Review by per1234 — CHANGES_REQUESTED** [review]
Since this is a big PR that contains multiple [atomic commits](http://www.freshconsulting.com/atomic-commits), I recommend it be granted an exemption from the repo's usual squash merge policy.

This would be done by temporarily adjusting [the configuration of the repo's merge button](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests), following this procedure:

1. Click the "**:gear: Settings**" tab of the repository that contains the PR
1. Under the "**Merge button**" section of the Settings page, check the box next to "**:black_square_button: Allow merge commits**"
1. Merge the PR, selecting the "**Merge pull request**" option from the merge button
1. Return to the repository's Settings page
1. Uncheck the box next to "**:black_square_button: Allow merge commits**"

Since the commits will not be automatically squashed, any fixup commits should be manually squashed via an [interactive rebase](https://git-scm.com/docs/git-rebase#_interactive_mode) before the merge.

File: cli/debug/debug.go
Comment (umbynos) [reply cli/debug/debug.go:61]: Actually I completely forgot to change this string... Thanks for noticing
Something like `Programmer to use, e.g: atmel_ice` should do?

Comment (per1234) [reply cli/debug/debug.go:61]: >Something like `Programmer to use, e.g: atmel_ice` should do?

Looks perfect!

File: cli/arguments/port.go
Comment (umbynos) [reply cli/arguments/port.go:45]: Putting everything in there was nor a smart idea, I'll create new files and move code around

File: cli/board/details.go
Comment (umbynos) [reply cli/board/details.go:66]: Done in 4d2df31, what do you think?

File: docs/UPGRADING.md
Comment (per1234) [inline docs/UPGRADING.md:15]:
```suggestion
The `board attach` command now uses `-p` and `-b` flags to respectively select the board FQBN and the port.
```

I think "uses" makes it more clear that this is a hard change rather than an additional syntax.

File: docs/UPGRADING.md
Comment (per1234) [inline docs/UPGRADING.md:18]:
```suggestion
The previous syntax `board attach <port>|<FQBN> [sketchPath]` is no longer supported.
```

File: docs/UPGRADING.md
Comment (per1234) [inline docs/UPGRADING.md:20]:
```suggestion
### `--timeout` flag in `board list` command has been replaced by `--discovery-timeout`
```
Minor grammar correction

File: docs/UPGRADING.md
Comment (per1234) [inline docs/UPGRADING.md:22]:
```suggestion
The flag `--timeout` in the `board list` command is no longer supported.
```

**Review by per1234 — CHANGES_REQUESTED** [review]

**Review by per1234 — APPROVED** [review]
Great stuff Umberto. Thanks!

File: cli/arguments/fqbn.go
Comment (silvanocerza) [inline cli/arguments/fqbn.go:43]:
```suggestion
func (f *Fqbn) Set(fqbn string) {
	f.fqbn = fqbn
}
```

File: cli/arguments/fqbn.go
Comment (silvanocerza) [inline cli/arguments/fqbn.go:38]:
```suggestion
func (f *Fqbn) String() string {
	return f.fqbn
}
```

File: cli/arguments/programmer.go
Comment (silvanocerza) [inline cli/arguments/programmer.go:38]:
```suggestion
func (p *Programmer) String() string {
	return p.programmer
}
```

File: cli/arguments/sketch.go
Comment (silvanocerza) [inline cli/arguments/sketch.go:56]:
```suggestion
func InitSketchPath(path string) *paths.Path {
```

We can always warn the users.

File: cli/arguments/sketch.go
Comment (silvanocerza) [inline cli/arguments/sketch.go:53]: Remove this function, it does way too many unrelated things.

File: cli/arguments/sketch.go
Comment (silvanocerza) [inline cli/arguments/sketch.go:94]: Try moving this in `Port` struct.

File: cli/arguments/sketch.go
Comment (silvanocerza) [inline cli/arguments/sketch.go:66]:
```suggestion
func NewSketch(sketchPath *paths.Path) *sketch.Sketch {
```

File: cli/instance/instance.go
Comment (silvanocerza) [inline cli/instance/instance.go:146]:
```suggestion
func CreateInstanceAndRunFirstUpdate() *rpc.Instance {
```

File: docs/UPGRADING.md
Comment (silvanocerza) [inline docs/UPGRADING.md:9]:
```suggestion
The `board details` command now accepts only the `--fqbn` or `-b` flags to specify the FQBN.
```

File: docs/UPGRADING.md
Comment (silvanocerza) [inline docs/UPGRADING.md:15]:
```suggestion
The `board attach` command now uses `--port` and `-p` flags to set board port and `--board` and `-b` flags to select its FQBN.
```

**Review by silvanocerza — CHANGES_REQUESTED** [review]
See comments.

**Review by silvanocerza — APPROVED** [review]

**Pull request merged**
